[ACR-1579] Add machine-learning-training docs

zhoonit opened the pull request:
Ready to be reviewed.

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
````diff
@@ -0,0 +1,152 @@
+   <feature name="http://tizen.org/feature/machine_learning">true</feature>
+   <feature name="http://tizen.org/feature/machine_learning.training">true</feature>
+   ```
+
````
add a statement about priveleges (if saving file outside the app, then apply privilege)?

zhoonit replied:
Done

zhoonit committed:
[ACR-1579] Add machine-learning-training docs

**Changes proposed in this PR:**
- Add machine-learning-training docs
- Update overview.md

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+# Machine Learning Training
+
+Machine Learning Training allows to construct, control and train a machine learning model in Tizen devices.
```
Machine Learning Training API/system/functionality?

zhoonit replied:
How about feature?

kparichay replied:
this will get mixed with Tizen features ().
Further, you use API at bottom. Make sure whatever you choose works with it.

kparichay replied:
Make it API, consistent wtih inference.

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+Please note that:
+
+* This feature remains experimental for Tizen 6.0 and is not full-fledged yet for every needs.
+* Every example code is not fully handling error for brevity. Error should be handled more extensively then example code written here.
```
not handling all the errors for brevity

kparichay replied:
then -> compared to the

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+Please note that:
+
+* This feature remains experimental for Tizen 6.0 and is not full-fledged yet for every needs.
```
Lets not add experimental. ACR reviewers were against it.
`This API is still undergoing heavy changes, and can be updated.`

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+## Prerequisites
+
+To enable your application to use the machine learning training functionality:
```
follow as per https://github.com/zhoonit/tizen-docs/pull/1/files#r478799057

zhoonit replied:
Did you mean this? 
```suggestion
To enable your application to use the machine learning training functionality, follow as per:
```

kparichay replied:
I mean machine learning training `functionality` -> as per decided from https://github.com/zhoonit/tizen-docs/pull/1/files#r478799057

zhoonit replied:
Changed to API

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+To enable your application to use the machine learning training functionality:
+
+1. To use the functions and data types of the Machine Learning Inference API, include the `<nntrainer.h>` header file in your application:
```
Inference -> Training

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+To enable your application to use the machine learning training functionality:
+
+1. To use the functions and data types of the Machine Learning Inference API, include the `<nntrainer.h>` header file in your application:
```
API -> as per  https://github.com/zhoonit/tizen-docs/pull/1/files#r478799057

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+To enable your application to use the machine learning training functionality:
+
+1. To use the functions and data types of the Machine Learning Inference API, include the `<nntrainer.h>` header file in your application:
```
Simply `include the `<nntrainer.h>` header file in your application:`. The first part of the sentence is already at line or modify line 12 to have it,.

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
````diff
@@ -0,0 +1,402 @@
+   #include <nntrainer.h>
+   ```
+
+2. To use the Machine Learning Training API, include the following features in your `tizen-manifest.xml` file:
````
ditto

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
````diff
@@ -0,0 +1,402 @@
+   <feature name="http://tizen.org/feature/machine_learning.training">true</feature>
+   ```
+
+3. If you want to save/load model files from outside, the application has to request permission by adding following privileges to the `tizen-manifest.xml` file:
````
outside `of the applicaiton's own resources`

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+## Building Blocks
+
+There are 4 major components you need to get familiar with. With those components, you can prepare a model and run.
```
run?? train etc?

zhoonit replied:
Changed to 
```suggestion
There are 4 major components you need to get familiar with. With those components, you can train a model.
```

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+![model](./media/model.png)
+
+Model is a wrapper component that has topology of layers, optimizers, datasets.
+Model conducts training and update parameters that can later be used for inference.
```
Model performs training and saves the updated parameters ....

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+ml_train_model_h model;
+ml_train_model_construct(&model);
+
+/* Configure models(omitted for brevity) */
```
models -> model

kparichay replied:
omitted for brevity -> detailed in later sections

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+### Optimizer
+
+Optimizer determines how to update model parameters according to loss from prediction.
+Currently, Stochastic gradient descent optimizer and Adam optimizer are supported.
```
Stochastic Gradient Descent

kparichay replied:
Currently -> Both

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+## Construct Model
+
+To run the model, you need to construct model. Model can be constructed with `ml_train_model_construct`.
```
you need to construct -> a model first needs to be construct.
Do similar across the content of the file.

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+## Construct Model
+
+To run the model, you need to construct model. Model can be constructed with `ml_train_model_construct`.
```
run the model -> what does run means - inference or train?

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+## Construct Model
+
+To run the model, you need to construct model. Model can be constructed with `ml_train_model_construct`.
+If you have file that describe model, the file can be used to construct initially with `ml_train_model_construct_with_file`.
```
describe -> describes

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+
+### Construct Model from the Description File
+
+As of now, only `*.ini` is supported to construct model from a file.
```
As of now, only INI formatted files `*.ini` are supported

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+#### Create Model from ini Format
+
+There are special sections refering to `model` and `dataset` respectively [Model] and [Dataset]
+Each ini section maps to a `layer`. Keys and values from each section sets property.
```
Rest of the ini sections map to `layer`.

kparichay replied:
Also add each layer name to unique.

zhoonit replied:
Done

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
````diff
@@ -0,0 +1,402 @@
+
+```ini
+[Model] # Special section that describes model property
+Type = NeuralNetwork	# Model Type : only NeuralNetwork is supported as of now
````
such information `only NeuralNetwork is supported as of now` must not be given here. But rather in docs of the project.

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+[outputlayer]
+Type = fully_connected
+Unit = 2		# Output Layer Dimension ( = Weight Width )
+bias_initializer = zeros
```
lets specify weight initializer as well for completeness

kparichay commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+There are some restriction that must be kept:
+
+ * Model file must have `[Model]` section that describe model property
+ * Model file must have at lease one layer.
```
at least

kparichay reviewed (commented):
I will review the rest and update this directly in the next PR.

jijoongmoon commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
```diff
@@ -0,0 +1,402 @@
+# Machine Learning Training
+
+Machine Learning Training allows to construct, control and train a machine learning model in Tizen devices.
```
I think we need to give little bit more details (briefly) about Machine Learning Training.
Something like..

Machine Learning Training API allows you to train your model in Tizen Devices.

The main features of the Machine Learning Training API include:

Construct deep neural network Easily
You can construct DNN model by Configuration File or writing code through Machine Learning Training API.
....
Train with your own data
Machine Learning API also allows you to train the model with your own data as a File I/O and defining data generator code.
....
Evaluate the model during train
You can validate and test your model during train process easily by define the data set.
...

zhoonit replied:
Updated

jijoongmoon commented on docs/application/native/guides/machine-learning/machine-learning-training.md:
````diff
@@ -0,0 +1,402 @@
+
+```
+/***************************************
+ * Data(x)  -> Layer(F) -> Data(F(X))  *
````
I think we don't need this. I think we need to give info about what kind of layer we can provide.

zhoonit replied:
Deleted
